doc/developer/design/20240609_error_handling.md

```diff
@@ -0,0 +1,416 @@
+
+## Solution Proposal
+
+### Motivation
+
+Only the Kafka source is particularly prone to bad data. With PostgreSQL and
+MySQL sources, it is almost always a bug (either in the upstream system or in
+Materialize) if data fails to decode, as the upstream systems enforce schemas.
+But Kafka does not guarantee schema enforcement, so it is common for bad data to
+slip into a Kafka topic.
+
+Because Kafka sources are the most prone to bad data, we propose to pursue a
+Kafka source-specific solution for handling source decoding errors.
+
+### User experience
+
+#### Dead letter queue
+
+We propose to introduce a [dead letter queue][dlq] (DLQ) for Kafka sources.
+
+Intuitively, the DLQ provides a running log of all decoding errors encountered
+by the source. Materialize users can monitor the count of errors in the DLQ and
+set up alerts whenever the count increases. Users can filter on `mz_timestamp`
+and/or `timestamp` to eliminate old errors that are no longer relevant. (Using
+the former column filters by ingestion time, while filtering on the latter
+column filters by the time the message was produced to Kafka.)
+
+Concretely, we propose to add a `REDIRECT ERRORS` option to `CREATE SOURCE` that
+specifies the name of a DLQ table in which to emit information about undecodable
+messages:
+
+```sql
+CREATE SOURCE src TO KAFKA CONNECTION kconn ... WITH (
+    REDIRECT ERRORS = dlq
+)
+```
+
+The specified DLQ table (`dlq` in the example above) must not exist before the
+`CREATE SOURCE` command is executed. Materialize will automatically create the
+DLQ table with the following Kafka-specific structure.
+
+ Name           | Type           | Nullable | Description
+----------------|----------------|----------|------------
+ `mz_timestamp` | `mz_timestamp` | No       | The logical timestamp that the message was reclocked to.
+ `error`        | `text`         | No       | The text of the decoding error.
+ `error_code`   | `text`         | No       | The code of the decoding error.
+ `partition`    | `integer`      | No       | The partition of the Kafka message that failed to decode.
+ `offset`       | `bigint`       | No       | The offset of the Kafka message that failed to decode.
+ `timestamp`    | `timestamp`    | No       | The timestamp of the Kafka message that failed to decode.
+ `key`          | `bytea`        | Yes      | The key bytes of the Kafka message that failed to decode.
+ `value`        | `bytea`        | Yes      | The value bytes of the Kafka message that failed to decode.
+
+The DLQ table works like a subsource. It can be queried via `SELECT`, subscribed
+to via `SUBSCRIBE`, and used as the upstream relation for a Kafka sink.
+
+The DLQ table is append only. Allowing users to specify a retention policy on
+the DLQ table is left to future work. Users cannot issue `INSERT`, `UPDATE`, or
+`DELETE` commands against the DLQ[^2].
+
+When a DLQ table is attached to a source, no errors are emitted to the source's
```

Comment:
How is a DLQ table attached to a source?

Reply:
Ah, sorry, when the `REDIRECT ERRORS` option is specified.